synapse/neuron.py

```diff
@@ -446,91 +445,182 @@ def initConfDefs(self):
-        self.txtinh = s_tinfoil.TinFoilHat(self.txkey)
+    def __init__(self, link, boss, lisn=False):
 
-        self.rxkey = None
-        self.rxtinh = None
+        s_net.Link.__init__(self, link)
+        self.chain(link)
+        self._sess_boss = boss
+        self.is_lisn = lisn    # True if we are the listener.
+        self._crypter = None  # type: CryptSeq
+        self._my_ephem_prv = None  # type: s_ecc.PriKey
 
     def handlers(self):
         return {
-            'cert': self._onMesgCert,
-            'skey': self._onMesgSkey,
+            'helo': self._onMesgHelo,
             'xmit': self._onMesgXmit,
+            'fail': self._onMesgFail
         }
 
-    def setRxKey(self, rxkey):
-        self.rxkey = rxkey
-        self.rxtinh = s_tinfoil.TinFoilHat(rxkey)
-
     def _tx_real(self, mesg):
 
-        if self.txtinh is None:
-            raise s_common.NotReady()
+        if self._crypter is None:
+            raise s_exc.NotReady()
 
-        data = self.txtinh.enc(s_msgpack.en(mesg))
+        data = self._crypter.encrypt(mesg)
         self.link.tx(('xmit', {'data': data}))
 
+    def _onMesgFail(self, link, mesg):
+        logger.error('Remote peer issued error: %r.', mesg)
+        self.txfini()
+
+    def _send_fail(self, exc):
+        self.link.tx(('fail', {'exception': repr(exc)}))
+
     def _onMesgXmit(self, link, mesg):
 
-        if self.rxtinh is None:
-            logger.warning('xmit message before rxkey')
+        if self._crypter is None:
+            logger.warning('xmit message before session establishment complete')
             raise s_common.NotReady()
 
-        data = mesg[1].get('data')
-        newm = s_msgpack.un(self.rxtinh.dec(data))
+        ciphertext = mesg[1].get('data')
+        try:
+            newm = self._crypter.decrypt(ciphertext)
+        except Exception as e:
+            self._send_fail(s_common.getexcfo(e))
+            logger.exception('decryption')
+            self.txfini()
+            return
 
         try:
             self.taskplex.rx(self, newm)
         except Exception as e:
+            self._send_fail(s_common.getexcfo(e))
             logger.exception('xmit taskplex error')
+            self.txfini()
 
-    def _onMesgSkey(self, link, mesg):
+    @s_glob.inpool
+    def _initiateSession(self):
+        ''' (As the initiator) start a new session
 
-        data = mesg[1].get('data')
+        Send ephemeral public and my certificate
+        '''
+        if self.is_lisn:
+            raise Exception('Listen link cannot initiate a session')
+        self._my_ephem_prv = s_ecc.PriKey.generate()
+        self.link.tx(('helo', {'version': NEURON_PROTO_VERSION,
+                               'ephem_pub': self._my_ephem_prv.public().dump(),
+                               'cert': self._sess_boss.certbyts}))
 
-        skey = self._sess_boss.decrypt(data)
+    def _handle_session_msg(self, mesg):
+        ''' Validate and set up the crypto from a helo message '''
```

Comment:
style nit.  We use new lines even for single line doc strings.

Reply:
ok